# Flight Operations Data Analysis




In [1]:
import pandas as pd

df = pd.read_csv("Day10_Flight_Operations_Dataset.csv")

df.head()


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5


In [2]:
print("Shape:", df.shape)

print("\nColumns:")
print(df.columns.tolist())

print("\nData types:")
print(df.dtypes)


Shape: (180, 17)

Columns:
['Flight_ID', 'Flight_Date', 'Airline', 'Origin', 'Destination', 'Aircraft', 'Travel_Class', 'Passengers', 'Seat_Capacity', 'Average_Ticket_Price', 'Delay_Minutes', 'Flight_Status', 'Weather', 'Booking_Channel', 'Avg_Baggage_Kg', 'Meal_Preference', 'Passenger_Satisfaction']

Data types:
Flight_ID                  object
Flight_Date                object
Airline                    object
Origin                     object
Destination                object
Aircraft                   object
Travel_Class               object
Passengers                  int64
Seat_Capacity               int64
Average_Ticket_Price        int64
Delay_Minutes             float64
Flight_Status              object
Weather                    object
Booking_Channel            object
Avg_Baggage_Kg              int64
Meal_Preference            object
Passenger_Satisfaction      int64
dtype: object


In [3]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180 entries, 0 to 179
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Flight_ID               180 non-null    object 
 1   Flight_Date             180 non-null    object 
 2   Airline                 180 non-null    object 
 3   Origin                  180 non-null    object 
 4   Destination             180 non-null    object 
 5   Aircraft                180 non-null    object 
 6   Travel_Class            180 non-null    object 
 7   Passengers              180 non-null    int64  
 8   Seat_Capacity           180 non-null    int64  
 9   Average_Ticket_Price    180 non-null    int64  
 10  Delay_Minutes           173 non-null    float64
 11  Flight_Status           180 non-null    object 
 12  Weather                 180 non-null    object 
 13  Booking_Channel         180 non-null    object 
 14  Avg_Baggage_Kg          180 non-null    in

In [4]:
print("First 5 rows:")
display(df.head())

print("Last 5 rows:")
display(df.tail())

print("Random 5 rows:")
display(df.sample(5))


First 5 rows:


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5


Last 5 rows:


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
175,FL0176,2026-04-10,IndiGo,Ahmedabad,Bengaluru,Airbus A320,Economy,48,189,4298,0.0,On Time,Storm,Mobile App,11,Vegetarian,5
176,FL0177,2026-02-17,Vistara,Chennai,Mumbai,Airbus A320,Business,124,180,17306,20.0,Delayed,Cloudy,Mobile App,8,No Meal,5
177,FL0178,2026-04-30,IndiGo,Delhi,Mumbai,Airbus A319,Economy,112,186,4490,110.0,Delayed,Fog,Online Travel Portal,16,Vegetarian,3
178,FL0179,2026-05-08,Vistara,Bengaluru,Kochi,Airbus A320,Economy,134,189,5413,20.0,Delayed,Rain,Online Travel Portal,9,No Meal,2
179,FL0180,2026-01-22,Akasa Air,Delhi,Mumbai,Airbus A321,Economy,144,210,3767,75.0,Delayed,Fog,Online Travel Portal,17,No Meal,5


Random 5 rows:


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
7,FL0008,2026-04-28,Air India,Delhi,Chennai,Airbus A319,Business,157,220,14173,0.0,On Time,Fog,Airline Website,25,Vegan,5
178,FL0179,2026-05-08,Vistara,Bengaluru,Kochi,Airbus A320,Economy,134,189,5413,20.0,Delayed,Rain,Online Travel Portal,9,No Meal,2
61,FL0062,2026-02-12,Air India,Bengaluru,Hyderabad,Airbus A319,Economy,131,210,6129,0.0,On Time,Clear,Travel Agency,7,Non-Vegetarian,3
57,FL0058,2026-04-14,IndiGo,Hyderabad,Kolkata,Boeing 737,Premium Economy,151,210,6834,110.0,Delayed,Cloudy,Travel Agency,16,Vegetarian,4
156,FL0157,2026-04-25,Akasa Air,Delhi,Srinagar,Boeing 737,Economy,120,186,2188,35.0,Delayed,Storm,Mobile App,14,Non-Vegetarian,2


In [5]:
## basic statistics
df.describe()


,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Avg_Baggage_Kg,Passenger_Satisfaction
count,180.000000,180.000000,180.000000,173.000000,180.000000,180.000000
mean,122.772222,197.238889,6111.677778,19.335260,15.150000,3.494444
std,45.635651,15.408176,3615.220040,29.029712,5.705506,1.169908
min,45.000000,180.000000,1949.000000,0.000000,5.000000,2.000000
25%,78.750000,186.000000,3815.500000,0.000000,10.000000,2.000000
50%,125.000000,189.000000,5380.000000,5.000000,15.500000,3.000000
75%,162.250000,210.000000,6790.000000,20.000000,20.000000,5.000000
max,210.000000,220.000000,27832.000000,110.000000,25.000000,5.000000


In [6]:
## Check Missing Values and Duplicates
print("Missing values:")
display(df.isnull().sum())

print("Duplicate rows:", df.duplicated().sum())


Missing values:


,0
Flight_ID,0
Flight_Date,0
Airline,0
Origin,0
Destination,0
Aircraft,0
Travel_Class,0
Passengers,0
Seat_Capacity,0
Average_Ticket_Price,0


Duplicate rows: 0


In [7]:
## `Delay_Minutes` contains missing values. These records are cancelled flights, so we will keep the original data unchanged and create a separate cleaned DataFrame only when doing delay-specific analysis.
df[df["Delay_Minutes"].isnull()][
    ["Flight_ID", "Flight_Date", "Flight_Status", "Delay_Minutes"]
]


,Flight_ID,Flight_Date,Flight_Status,Delay_Minutes
23,FL0024,2026-05-19,Cancelled,NaN
37,FL0038,2026-03-27,Cancelled,NaN
40,FL0041,2026-03-23,Cancelled,NaN
65,FL0066,2026-03-06,Cancelled,NaN
68,FL0069,2026-03-09,Cancelled,NaN
110,FL0111,2026-05-16,Cancelled,NaN
160,FL0161,2026-04-03,Cancelled,NaN


In [8]:
## Convert the date column to datetime
df["Flight_Date"] = pd.to_datetime(df["Flight_Date"])

df[["Flight_ID", "Flight_Date"]].head()


,Flight_ID,Flight_Date
0,FL0001,2026-03-23
1,FL0002,2026-06-07
2,FL0003,2026-05-12
3,FL0004,2026-05-30
4,FL0005,2026-02-27


In [9]:
# Single column
df["Airline"].head()


,Airline
0,SpiceJet
1,Akasa Air
2,Air India
3,Akasa Air
4,Air India


In [10]:
# Multiple columns
df[[
    "Flight_ID",
    "Airline",
    "Origin",
    "Destination",
    "Flight_Status"
]].head(10)


,Flight_ID,Airline,Origin,Destination,Flight_Status
0,FL0001,SpiceJet,Mumbai,Bengaluru,Delayed
1,FL0002,Akasa Air,Delhi,Mumbai,On Time
2,FL0003,Air India,Kochi,Delhi,On Time
3,FL0004,Akasa Air,Srinagar,Delhi,Delayed
4,FL0005,Air India,Pune,Delhi,Delayed
5,FL0006,Akasa Air,Kochi,Delhi,Delayed
6,FL0007,Air India Express,Delhi,Srinagar,On Time
7,FL0008,Air India,Delhi,Chennai,On Time
8,FL0009,Akasa Air,Delhi,Chennai,Delayed
9,FL0010,Akasa Air,Ahmedabad,Bengaluru,On Time


In [11]:
# loc: select specific rows and named columns
df.loc[0:4, [
    "Flight_ID",
    "Airline",
    "Passengers",
    "Flight_Status"
]]


,Flight_ID,Airline,Passengers,Flight_Status
0,FL0001,SpiceJet,210,Delayed
1,FL0002,Akasa Air,70,On Time
2,FL0003,Air India,67,On Time
3,FL0004,Akasa Air,153,Delayed
4,FL0005,Air India,178,Delayed


In [12]:
# iloc: select rows and columns by position
df.iloc[0:5, 0:6]


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319


##  Filtering



### Flights delayed by more than 30 minutes


In [13]:
df[df["Delay_Minutes"] > 30][[
    "Flight_ID",
    "Airline",
    "Origin",
    "Destination",
    "Delay_Minutes"
]]


,Flight_ID,Airline,Origin,Destination,Delay_Minutes
0,FL0001,SpiceJet,Mumbai,Bengaluru,50.0
3,FL0004,Akasa Air,Srinagar,Delhi,35.0
5,FL0006,Akasa Air,Kochi,Delhi,35.0
8,FL0009,Akasa Air,Delhi,Chennai,75.0
21,FL0022,Vistara,Kochi,Delhi,75.0
26,FL0027,SpiceJet,Bengaluru,Hyderabad,110.0
29,FL0030,SpiceJet,Delhi,Srinagar,75.0
30,FL0031,Vistara,Delhi,Srinagar,50.0
32,FL0033,Akasa Air,Ahmedabad,Bengaluru,35.0
42,FL0043,Vistara,Delhi,Chennai,50.0


### Flights that are delayed and have passenger satisfaction of at least 4


In [14]:
df[
    (df["Flight_Status"] == "Delayed") &
    (df["Passenger_Satisfaction"] >= 4)
][[
    "Flight_ID",
    "Airline",
    "Delay_Minutes",
    "Passenger_Satisfaction"
]].head(15)


,Flight_ID,Airline,Delay_Minutes,Passenger_Satisfaction
4,FL0005,Air India,20.0,5
8,FL0009,Akasa Air,75.0,4
17,FL0018,Air India Express,20.0,4
26,FL0027,SpiceJet,110.0,4
30,FL0031,Vistara,50.0,4
32,FL0033,Akasa Air,35.0,4
47,FL0048,Air India,20.0,5
50,FL0051,Air India,110.0,4
57,FL0058,IndiGo,110.0,4
69,FL0070,SpiceJet,20.0,4


### Flights from Delhi or Mumbai


In [15]:
df[
    (df["Origin"] == "Delhi") |
    (df["Origin"] == "Mumbai")
][[
    "Flight_ID",
    "Origin",
    "Destination",
    "Airline"
]].head(15)


,Flight_ID,Origin,Destination,Airline
0,FL0001,Mumbai,Bengaluru,SpiceJet
1,FL0002,Delhi,Mumbai,Akasa Air
6,FL0007,Delhi,Srinagar,Air India Express
7,FL0008,Delhi,Chennai,Air India
8,FL0009,Delhi,Chennai,Akasa Air
10,FL0011,Mumbai,Bengaluru,Air India
15,FL0016,Mumbai,Bengaluru,Akasa Air
16,FL0017,Delhi,Chennai,SpiceJet
19,FL0020,Mumbai,Kolkata,SpiceJet
22,FL0023,Delhi,Mumbai,Akasa Air


##  Sorting



In [16]:
df.sort_values(
    "Delay_Minutes",
    ascending=False
)[[
    "Flight_ID",
    "Airline",
    "Delay_Minutes",
    "Flight_Status"
]].head(10)


,Flight_ID,Airline,Delay_Minutes,Flight_Status
26,FL0027,SpiceJet,110.0,Delayed
177,FL0178,IndiGo,110.0,Delayed
82,FL0083,Air India Express,110.0,Delayed
57,FL0058,IndiGo,110.0,Delayed
50,FL0051,Air India,110.0,Delayed
128,FL0129,SpiceJet,110.0,Delayed
146,FL0147,IndiGo,110.0,Delayed
169,FL0170,IndiGo,110.0,Delayed
161,FL0162,Air India,110.0,Delayed
21,FL0022,Vistara,75.0,Delayed


In [17]:
df.sort_values(
    "Average_Ticket_Price",
    ascending=False
)[[
    "Flight_ID",
    "Airline",
    "Travel_Class",
    "Average_Ticket_Price"
]].head(10)


,Flight_ID,Airline,Travel_Class,Average_Ticket_Price
138,FL0139,Air India Express,Business,27832
52,FL0053,Akasa Air,Business,21499
32,FL0033,Akasa Air,Business,17625
176,FL0177,Vistara,Business,17306
108,FL0109,SpiceJet,Business,16390
132,FL0133,IndiGo,Business,15324
155,FL0156,Vistara,Business,15040
7,FL0008,Air India,Business,14173
51,FL0052,SpiceJet,Business,13824
92,FL0093,Air India Express,Business,12970


## Clean Delay Data for Delay-Specific Analysis




In [18]:
delay_df = df.dropna(subset=["Delay_Minutes"]).copy()

print("Original rows:", len(df))
print("Rows with recorded delay:", len(delay_df))
print("Rows removed for delay analysis:", len(df) - len(delay_df))


Original rows: 180
Rows with recorded delay: 173
Rows removed for delay analysis: 7


## Transform — Create Useful Columns



In [19]:
# Extract month from the datetime column
df["Flight_Month"] = df["Flight_Date"].dt.month

# Occupancy rate
df["Occupancy_Rate"] = (
    df["Passengers"] / df["Seat_Capacity"] * 100
).round(2)

df[[
    "Flight_ID",
    "Flight_Date",
    "Flight_Month",
    "Passengers",
    "Seat_Capacity",
    "Occupancy_Rate"
]].head(10)


,Flight_ID,Flight_Date,Flight_Month,Passengers,Seat_Capacity,Occupancy_Rate
0,FL0001,2026-03-23,3,210,180,116.67
1,FL0002,2026-06-07,6,70,210,33.33
2,FL0003,2026-05-12,5,67,220,30.45
3,FL0004,2026-05-30,5,153,220,69.55
4,FL0005,2026-02-27,2,178,180,98.89
5,FL0006,2026-05-20,5,123,220,55.91
6,FL0007,2026-06-17,6,178,210,84.76
7,FL0008,2026-04-28,4,157,220,71.36
8,FL0009,2026-01-25,1,82,210,39.05
9,FL0010,2026-06-12,6,125,220,56.82


### Create a delay category using `apply()`




In [20]:
def delay_category(minutes):
    if pd.isna(minutes):
        return "Not Recorded"
    elif minutes == 0:
        return "On Time"
    elif minutes <= 30:
        return "Short Delay"
    elif minutes <= 60:
        return "Long Delay"
    else:
        return "Very Long Delay"

df["Delay_Category"] = df["Delay_Minutes"].apply(delay_category)

df[[
    "Flight_ID",
    "Delay_Minutes",
    "Delay_Category"
]].head(15)


,Flight_ID,Delay_Minutes,Delay_Category
0,FL0001,50.0,Long Delay
1,FL0002,0.0,On Time
2,FL0003,5.0,Short Delay
3,FL0004,35.0,Long Delay
4,FL0005,20.0,Short Delay
5,FL0006,35.0,Long Delay
6,FL0007,0.0,On Time
7,FL0008,0.0,On Time
8,FL0009,75.0,Very Long Delay
9,FL0010,12.0,Short Delay


### Use `map()` for a simple transformation




In [21]:
status_priority = {
    "On Time": 1,
    "Delayed": 2,
    "Cancelled": 3
}

df["Status_Priority"] = df["Flight_Status"].map(status_priority)

df[[
    "Flight_ID",
    "Flight_Status",
    "Status_Priority"
]].head(10)


,Flight_ID,Flight_Status,Status_Priority
0,FL0001,Delayed,2
1,FL0002,On Time,1
2,FL0003,On Time,1
3,FL0004,Delayed,2
4,FL0005,Delayed,2
5,FL0006,Delayed,2
6,FL0007,On Time,1
7,FL0008,On Time,1
8,FL0009,Delayed,2
9,FL0010,On Time,1


##  Analyse — Grouping and Aggregation




In [22]:
airline_summary = df.groupby("Airline").agg({
    "Flight_ID": "count",
    "Passengers": "sum",
    "Average_Ticket_Price": "mean",
    "Passenger_Satisfaction": "mean"
})

airline_summary = airline_summary.rename(columns={
    "Flight_ID": "Number_of_Flights",
    "Passengers": "Total_Passengers",
    "Average_Ticket_Price": "Average_Ticket_Price",
    "Passenger_Satisfaction": "Average_Satisfaction"
})

airline_summary.round(2)


,Number_of_Flights,Total_Passengers,Average_Ticket_Price,Average_Satisfaction
Airline,,,,
Air India,24,3074,5083.08,3.75
Air India Express,24,3027,7388.46,3.67
Akasa Air,31,3427,5841.00,3.42
IndiGo,32,4158,5931.66,3.34
SpiceJet,32,3929,6189.75,3.53
Vistara,37,4484,6265.65,3.38


### Average delay by airline




In [23]:
airline_delay = delay_df.groupby("Airline")["Delay_Minutes"].mean()

airline_delay = airline_delay.rename("Average_Delay")

airline_delay.sort_values(ascending=False).round(2)


,Average_Delay
Airline,
IndiGo,25.03
SpiceJet,24.14
Air India,21.26
Vistara,18.63
Akasa Air,14.40
Air India Express,11.29


## Flight Status Analysis


In [24]:
status_summary = df.groupby("Flight_Status")["Flight_ID"].count()
status_summary


,Flight_ID
Flight_Status,
Cancelled,7
Delayed,59
On Time,114


In [25]:
status_percentage = (
    df["Flight_Status"].value_counts(normalize=True) * 100
).round(2)

status_percentage


,proportion
Flight_Status,
On Time,63.33
Delayed,32.78
Cancelled,3.89


##  Weather and Delay Analysis


In [26]:
weather_summary = delay_df.groupby("Weather").agg({
    "Flight_ID": "count",
    "Delay_Minutes": "mean"
}).rename(columns={
    "Flight_ID": "Flights_With_Recorded_Delay",
    "Delay_Minutes": "Average_Delay"
})

weather_summary.sort_values(
    "Average_Delay",
    ascending=False
).round(2)


,Flights_With_Recorded_Delay,Average_Delay
Weather,,
Rain,41,22.22
Storm,29,20.10
Cloudy,35,20.03
Clear,34,18.03
Fog,34,15.79


## Travel Class Analysis


In [27]:
class_summary = df.groupby("Travel_Class").agg({
    "Flight_ID": "count",
    "Average_Ticket_Price": "mean",
    "Passenger_Satisfaction": "mean"
}).rename(columns={
    "Flight_ID": "Number_of_Flights",
    "Average_Ticket_Price": "Average_Ticket_Price",
    "Passenger_Satisfaction": "Average_Satisfaction"
})

class_summary.round(2)


,Number_of_Flights,Average_Ticket_Price,Average_Satisfaction
Travel_Class,,,
Business,11,16780.00,3.27
Economy,134,4550.27,3.46
Premium Economy,35,8736.74,3.69


##  Monthly Analysis


In [28]:
monthly_summary = df.groupby("Flight_Month").agg({
    "Flight_ID": "count",
    "Passengers": "sum"
}).rename(columns={
    "Flight_ID": "Number_of_Flights",
    "Passengers": "Total_Passengers"
})

monthly_summary


,Number_of_Flights,Total_Passengers
Flight_Month,,
1,29,3656
2,28,3502
3,33,3839
4,35,4354
5,30,3504
6,25,3244


##  Route Analysis




In [29]:
route_summary = df.groupby(
    ["Origin", "Destination"]
).agg({
    "Flight_ID": "count",
    "Passengers": "sum"
}).rename(columns={
    "Flight_ID": "Number_of_Flights",
    "Passengers": "Total_Passengers"
})

route_summary.sort_values(
    "Number_of_Flights",
    ascending=False
).head(10)


Number_of_Flights  Total_Passengers
Origin    Destination                                     
Delhi     Chennai                     17              2011
Mumbai    Bengaluru                   16              1933
Srinagar  Delhi                       15              1957
Kochi     Delhi                       14              1657
Delhi     Srinagar                    13              1828
          Mumbai                      13              1609
Bengaluru Hyderabad                   13              1795
Mumbai    Kolkata                     13              1683
Ahmedabad Bengaluru                   12              1411
Lucknow   Delhi                       11              1137

## Transformation with `melt()`




In [30]:
airline_metrics = airline_summary.reset_index()[[
    "Airline",
    "Average_Ticket_Price",
    "Average_Satisfaction"
]]

airline_metrics_long = pd.melt(
    airline_metrics,
    id_vars=["Airline"],
    var_name="Metric",
    value_name="Value"
)

airline_metrics_long.head(12)


,Airline,Metric,Value
0,Air India,Average_Ticket_Price,5083.083333
1,Air India Express,Average_Ticket_Price,7388.458333
2,Akasa Air,Average_Ticket_Price,5841.000000
3,IndiGo,Average_Ticket_Price,5931.656250
4,SpiceJet,Average_Ticket_Price,6189.750000
5,Vistara,Average_Ticket_Price,6265.648649
6,Air India,Average_Satisfaction,3.750000
7,Air India Express,Average_Satisfaction,3.666667
8,Akasa Air,Average_Satisfaction,3.419355
9,IndiGo,Average_Satisfaction,3.343750


##  Combining Data with `concat()`



In [31]:
first_half = df.iloc[:90]
second_half = df.iloc[90:]

combined_df = pd.concat(
    [first_half, second_half],
    ignore_index=True
)

print("Original rows:", len(df))
print("Rows after concat:", len(combined_df))

combined_df.head()


Original rows: 180
Rows after concat: 180


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,...,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Flight_Month,Occupancy_Rate,Delay_Category,Status_Priority
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,...,Delayed,Storm,Online Travel Portal,23,No Meal,3,3,116.67,Long Delay,2
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,...,On Time,Fog,Travel Agency,16,Vegan,5,6,33.33,On Time,1
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,...,On Time,Rain,Travel Agency,19,No Meal,2,5,30.45,Short Delay,1
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,...,Delayed,Storm,Travel Agency,18,No Meal,2,5,69.55,Long Delay,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,...,Delayed,Cloudy,Airline Website,25,No Meal,5,2,98.89,Short Delay,2


##  Combining Data with `merge()`




In [32]:
passenger_summary = (
    df.groupby("Airline")["Passengers"]
    .sum()
    .reset_index()
    .rename(columns={"Passengers": "Total_Passengers"})
)

delay_summary = (
    delay_df.groupby("Airline")["Delay_Minutes"]
    .mean()
    .reset_index()
    .rename(columns={"Delay_Minutes": "Average_Delay"})
)

merged_airline = pd.merge(
    passenger_summary,
    delay_summary,
    on="Airline"
)

merged_airline.sort_values(
    "Total_Passengers",
    ascending=False
).round(2)


,Airline,Total_Passengers,Average_Delay
5,Vistara,4484,18.63
3,IndiGo,4158,25.03
4,SpiceJet,3929,24.14
2,Akasa Air,3427,14.40
0,Air India,3074,21.26
1,Air India Express,3027,11.29


##  Data Type and Memory Optimization




In [33]:
df.memory_usage(deep=True)


,0
Index,132
Flight_ID,9900
Flight_Date,1440
Airline,10430
Origin,9991
Destination,10027
Aircraft,10756
Travel_Class,10371
Passengers,1440
Seat_Capacity,1440


### Convert repeated text columns to `category`




In [34]:
optimized_df = df.copy()

category_columns = [
    "Airline",
    "Origin",
    "Destination",
    "Aircraft",
    "Travel_Class",
    "Flight_Status",
    "Weather",
    "Booking_Channel",
    "Meal_Preference",
    "Delay_Category"
]

for col in category_columns:
    optimized_df[col] = optimized_df[col].astype("category")

optimized_df.memory_usage(deep=True)


,0
Index,132
Flight_ID,9900
Flight_Date,1440
Airline,702
Origin,1094
Destination,928
Aircraft,591
Travel_Class,465
Passengers,1440
Seat_Capacity,1440


### Numeric downcasting




In [35]:
for col in ["Passengers", "Seat_Capacity", "Average_Ticket_Price", "Avg_Baggage_Kg", "Passenger_Satisfaction"]:
    optimized_df[col] = pd.to_numeric(
        optimized_df[col],
        downcast="integer"
    )

optimized_df.dtypes


,0
Flight_ID,object
Flight_Date,datetime64[ns]
Airline,category
Origin,category
Destination,category
Aircraft,category
Travel_Class,category
Passengers,int16
Seat_Capacity,int16
Average_Ticket_Price,int16


##  Data Quality Check



In [36]:
capacity_issues = df[df["Passengers"] > df["Seat_Capacity"]][[
    "Flight_ID",
    "Airline",
    "Passengers",
    "Seat_Capacity",
    "Occupancy_Rate"
]]

print("Records with occupancy above 100%:", len(capacity_issues))
display(capacity_issues)


Records with occupancy above 100%: 9


,Flight_ID,Airline,Passengers,Seat_Capacity,Occupancy_Rate
0,FL0001,SpiceJet,210,180,116.67
10,FL0011,Air India,195,189,103.17
30,FL0031,Vistara,209,180,116.11
69,FL0070,SpiceJet,201,189,106.35
80,FL0081,SpiceJet,199,180,110.56
94,FL0095,Vistara,199,189,105.29
107,FL0108,IndiGo,198,180,110.00
126,FL0127,IndiGo,184,180,102.22
153,FL0154,Air India,201,186,108.06


##  Answering Meaningful Questions

### Which airline has the most flight records?


In [37]:
airline_summary.sort_values(
    "Number_of_Flights",
    ascending=False
).head(1)


,Number_of_Flights,Total_Passengers,Average_Ticket_Price,Average_Satisfaction
Airline,,,,
Vistara,37,4484,6265.648649,3.378378


### Which airline has the highest average recorded delay?


In [38]:
airline_delay.sort_values(
    ascending=False
).head(1).round(2)


,Average_Delay
Airline,
IndiGo,25.03


### Which weather condition has the highest average recorded delay?


In [ ]:
weather_summary.sort_values(
    "Average_Delay",
    ascending=False
).head(1).round(2)


### Which travel class has the highest passenger satisfaction?


In [39]:
class_summary.sort_values(
    "Average_Satisfaction",
    ascending=False
).head(1).round(2)


,Number_of_Flights,Average_Ticket_Price,Average_Satisfaction
Travel_Class,,,
Premium Economy,35,8736.74,3.69


### Which route appears most frequently?


In [40]:
route_summary.sort_values(
    "Number_of_Flights",
    ascending=False
).head(1)


,,Number_of_Flights,Total_Passengers
Origin,Destination,,
Delhi,Chennai,17,2011


# Key Observations

1. **Most flights are On Time:** 114 out of 180 flights are On Time, while 59 are Delayed and 7 are Cancelled.

2. **Vistara has the most flight records:** Vistara appears in 37 of the 180 records, the highest among the airlines.

3. **IndiGo has the highest average recorded delay:** Its average recorded delay is about **25.03 minutes**, while Air India Express has the lowest at about **11.29 minutes**.

4. **Rain has the highest average recorded delay:** Among flights with a recorded delay, Rain has an average delay of about **22.22 minutes**.

5. **Premium Economy has the highest average passenger satisfaction:** Its average satisfaction is about **3.69/5**, higher than Economy and Business.

6. **Online Travel Portal is the most common booking channel:** It appears in **50 flights**, followed by Mobile App with 45.

7. **Delhi–Chennai is the most frequently recorded route:** It appears **17 times** in the dataset.

8. **The dataset contains data-quality issues:** 7 flights have missing delay values because they are cancelled, and 9 records have passengers greater than the recorded seat capacity.


#  Export

The final transformed DataFrame can be exported as a CSV using `to_csv()`.


In [41]:
output_path = "Flight_Operations_Analysed.csv"

df.to_csv(output_path, index=False)

print("Exported successfully to:", output_path)


Exported successfully to: Flight_Operations_Analysed.csv
